src/INode.java

```diff
@@ -0,0 +1,68 @@
+import java.util.ArrayList;
+
+/**
+ * An abstract syntax tree internal node.
+ */
+public class INode extends AST {
+	public ArrayList<AST> children;
+	
+	public INode(Type type, Object... children) {
+		super(type);
+		
+		this.children = new ArrayList<AST>(children.length);
+		
+		for (int i = 0; i < children.length; i++)
+			if (children[i] == null)
+				this.children.add(null);
+			else
+				this.children.add((AST) children[i]);
+	}
+	
+	/**
+	 * Appends a given list of nodes to this node's list of children. 
+	 * 
+	 * @param nodes the list of nodes to be added
+	 * @return this node
+	 */
+	public INode add(ArrayList<AST> nodes) {
+		for (AST node : nodes)
+			this.children.add(node);
+		
+		return this;
+	}
+	
+	private void indent(int level) {
+		for (int i = 0; i < level; i++)
+			System.err.printf("  ");		
+	}
+	
+	public void dump(int level) {
+		indent(level);
+		System.err.println(this.type.toString());
+		
+		if (children.size() > 0) {
+			for (AST child : children) {
+				if (child != null) {
+					child.dump(level + 1);
+				}
+			}
+		} else {
+			indent(level + 1);
+			System.err.println("  empty");
+		}		
+	}
+	
+	@Override
+	public String toString() {
+		String s;
+		String c = "";
```

Comment:
Should be a `StringBuffer` instead of a `String`. A string is a set length of memory in Java, so a new length of memory will need to be allocated and all chars moved each time you concatenate strings. A StringBuilder, on which you call `.toString()` at the end, doesn't need to copy each character on concatenation.

Not strictly a necessary change (since this is fully working and I doubt we're being graded at all on our Java String efficiency), but I figure it's best if we give each other additional feedback on our code so that we can each improve our Java programming.